src/magic_folder/downloader.py

```diff
@@ -372,7 +379,14 @@ def _process(self, snapshot):
                     name=snapshot.name,
                     capability=snapshot.capability,
                 ):
-                    staged = yield self._magic_fs.download_content_to_staging(snapshot, self.tahoe_client)
+                    try:
+                        staged = yield self._magic_fs.download_content_to_staging(snapshot, self.tahoe_client)
+                    except Exception:
+                        raise PublicError(
+                            self._clock.seconds(),
+                            "Failed to download snapshot for '{}'.".format(relpath)
+                        )
+
```

Comment:
My first inclination would be that it is would be better directly report the error here (by calling `.error_occured` and then re-raising the original exception. After thinking about it further, I think that makes sense, as we may not want to let this error propagate to the top-level; in this particular case, we may want to continue scanning the collective, even if we can't update one file.

As an aside, I think this case differs from the `APIError` case, in that in the `APIError` case, there is *by definition* a caller that will handle it, when relevant. That isn't to say that every function that raises it is only called from the API, but that when we may want to report an error via the API, we are being called from there, and so there is something that catches `APIError`.

In this case, there isn't such a top-level handler, so we would need to add them as appropriate, and the places we may need those will grow as improve the software to partially recover from errors.

If we make this change, that also means that we can leave getting the timestamp to the status API side, so we don't need to pass in a reactor here.

Reply:
I did have it more like that at first, but then it's like "`error_occurred(PublicError(..)); raise`" most of the time .. and was thinking with a top-level handler we can then raise `PublicError` directly more places (if the need arises -- I'm sure there will be more errors we want to report to "users").

time-stamp-wise, sure maybe that's better just handled directly by `IStatus` instead of at each raise-point.

As far as "continue scanning, or not" this code doesn't change that behavior (before, it would leak the exception out anyway). I'm not sure if there _is_ a right answer here for that .. some errors are probably "per-file" but if it's like "disc full" or "can't write the magic-dir" then you're going to flood a bunch of errors on every scan, basically?

So, what if I take the timestamp out of the `PublicError` ctor (and make the status service insert them instead) and we can file a ticket about (or otherwise address) the wider issue of "do we keep scanning on errors in a folder, or not"..?

Reply:
Regarding the handling: I did add a higher-level handler (at the "downloader service" level).

Reply:
> I did have it more like that at first, but then it's like "error_occurred(PublicError(..)); raise" most of the time

That seems like a reasonable thing to have (though it isn't clear to me that the `PublicError` there adds anything). That is exactly the interface I was suggesting.

> with a top-level handler we can then raise PublicError directly more places

I'm not sure what you mean by this? I guess `.snapshots` and `.participants` don't have a status instance to call, maybe? My guess is that those places may not have enough context for a good error message, and anywhere else can report a status error directly as easily as raise `PublicError`.

> As far as "continue scanning, or not" this code doesn't change that behavior

Sure, this was thinking about how we might change the code, and why we might not want to depend on `PublicError` escaping to a particular exception handler. I wasn't suggesting changing anything, just thinking through what impacts the API we pick here would be, if we made that change.

> Regarding the handling: I did add a higher-level handler (at the "downloader service" level).

Yes. My point about possibly handling errors is that we may decide we want to handle errors before they get to there, which may lead to a proliferation of places where we need to catch `PublicError` and report it, in *addition* to all the places where we raise it.

Reply:
> So, what if I take the timestamp out of the PublicError ctor (and make the status service insert them instead)

To be clear, I think we should go with the API that does not use exceptions.